python/pyspark/sql/tests/test_udtf.py

```diff
@@ -0,0 +1,366 @@
+        @udtf(returnType="key: string, value: float")
+        class TestUDTF:
+            def __init__(self):
+                self._count = 0
+                self._sum = 0
+
+            def eval(self, x: int):
+                if x > 0:
+                    self._count += 1
+                    self._sum += x
+                    yield "input", float(x)
+
+            def terminate(self):
+                yield "count", float(self._count)
+                yield "avg", self._sum / self._count
+
+        self.assertEqual(
+            TestUDTF(lit(1)).collect(),
+            [Row(key="input", value=1), Row(key="count", value=1.0), Row(key="avg", value=1.0)],
+        )
+
+        with self.assertRaisesRegex(PythonException, "division by zero"):
+            TestUDTF(lit(0)).collect()
+
+        self.spark.udtf.register("test_udtf", TestUDTF)
+        df = self.spark.sql(
+            "SELECT id, key, value FROM range(0, 10, 1, 2), "
+            "LATERAL test_udtf(id) WHERE key != 'input'"
+        )
+        self.assertEqual(
+            df.collect(),
+            [
+                Row(id=0, key="count", value=4.0),
+                Row(id=0, key="avg", value=2.5),
+                Row(id=0, key="count", value=5.0),
+                Row(id=0, key="avg", value=7.0),
+            ],
+        )
+
+    def test_terminate_with_exceptions(self):
+        @udtf(returnType="a: int, b: int")
+        class TestUDTF:
+            def eval(self, a: int):
+                yield a, a + 1
+
+            def terminate(self):
+                raise ValueError("terminate error")
+
+        with self.assertRaisesRegex(
+            PythonException, "Failed to terminate the user defined table function: terminate error"
+        ):
+            TestUDTF(lit(1)).collect()
+
+    def test_nondeterministic_udtf(self):
+        import random
+
+        class RandomUDTF:
+            def eval(self, a: int):
+                yield a * int(random.random() * 100),
+
+        random_udtf = udtf(RandomUDTF, returnType="x: int").asNondeterministic()
+        # TODO(SPARK-43966): support non-deterministic UDTFs
+        with self.assertRaisesRegex(AnalysisException, "nondeterministic expressions"):
+            random_udtf(lit(1)).collect()
+
+    def test_udtf_with_nondeterministic_input(self):
+        from pyspark.sql.functions import rand
+
+        @udtf(returnType="x: int")
+        class TestUDTF:
+            def eval(self, a: int):
+                yield a + 1,
+
+        # TODO(SPARK-43966): support non-deterministic UDTFs
+        with self.assertRaisesRegex(AnalysisException, "nondeterministic expressions"):
+            TestUDTF(rand(0) * 100).collect()
+
+    def test_udtf_no_eval(self):
+        @udtf(returnType="a: int, b: int")
+        class TestUDTF:
+            def run(self, a: int):
+                yield a, a + 1
+
+        with self.assertRaisesRegex(PythonException, "Python UDTF must implement the eval method"):
```

Comment:
Same, can we include the UDTF name in the error message, e.g.

```
Failed to execute the user defined table function TestUDTF because
it has not yet implemented the "eval" method; please add the "eval"
method and try the query again
```